### PR TITLE
Xeps/bind2

### PR DESCRIPTION
**Pull request by NelsonVides**
Implements partial support for bind2. Here we can already bind a resource and start a session directly after authentication, saving one more round-trip.

Support for carbons and csi on the same roundtrip coming in the next PR #4114

### CODE REVIEW

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/189956) / elasticsearch_and_cassandra_mnesia / 26e64da756880cf37cf725a20e7052afb65498f6
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4113/189956/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4113/189956/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%409ca79bcdf06d.2023-09-04_14.33.10/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190089) / elasticsearch_and_cassandra_mnesia / b7b3a53a35d22376ae09d2b67df86c1c2e578c97
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4113/190089/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4113/190089/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40a1475c19c9be.2023-09-05_07.57.48/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190155) / elasticsearch_and_cassandra_mnesia / 9ee0ac33310f23a66452de909193d681f2f2c253
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4113/190155/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4113/190155/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40b3b97b908044.2023-09-05_11.20.03/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/190208) / elasticsearch_and_cassandra_mnesia / c460ef39166bc5c91b85e5e671ec6b60e281885b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4113/190208/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4113/190208/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40f0edc6b0a397.2023-09-05_15.57.27/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

File: src/mod_bind2.erl [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            NewStateData = SaslStateData#{c2s_data => C2SData, c2s_state => session_established},
+            SaslAcc2 = mod_sasl2:set_state_data(SaslAcc1, NewStateData),
+            SaslAcc3 = mod_sasl2:update_inline_request(SaslAcc2, ?MODULE, Bound, success),
+            SaslAcc4 = mongoose_hooks:bind2_enable_features(HostType, SaslAcc3, SaslStateData),
+            {ok, SaslAcc4};
+        {stop, SaslAcc1} ->
+            %% The XEP does not specify what to do if the resource wasn't bound
+            %% so we just set failed here and move along with SASL2
+            Error = #xmlel{name = <<"failed">>, attrs = [?XMLNS_BIND2], children = MaybeSmChild},
+            {ok, mod_sasl2:update_inline_request(SaslAcc1, ?MODULE, Error, failure)}
```
Comment (arcusfelis) [inline src/mod_bind2.erl:126]: would be nice to have a test here (not covered)

**Review by arcusfelis — APPROVED** [review]
ok

**Comment by Neustradamus** [comment]
@NelsonVides: Good job!

**Pull request merged**